storage/innobase/trx/trx0trx.cc

```diff
@@ -1260,20 +1260,33 @@ static void trx_flush_log_if_needed(lsn_t lsn, trx_t *trx)
   const bool flush=
     (srv_file_flush_method != SRV_NOSYNC &&
      (srv_flush_log_at_trx_commit & 1));
-
-  completion_callback cb;
-  if (!log_sys.is_pmem() &&
-      (cb.m_param= thd_increment_pending_ops(trx->mysql_thd)))
-  {
-    cb.m_callback = (void (*)(void *)) thd_decrement_pending_ops;
-    log_write_up_to(lsn, flush, &cb);
-  }
+#if defined HAVE_INNODB_MMAP && defined __linux__
+  if (!log_sys.is_mmap());
+  else if (!flush)
+    /* Starting with Linux 2.6.19, msync(MS_ASYNC) is a no-op
+    and we do not need to do anything special for non-durable writes. */
+    return;
```

Comment:
We will need such logic also on other platforms that implement memory-mapped log files.

An assertion is failing at the start of `log_write_up_to()` [on Windows](https://buildbot.mariadb.org/#/builders/234/builds/31540/steps/9/logs/stdio):
```
CURRENT_TEST: mariabackup.undo_upgrade
mysqltest: At line 11: query 'CREATE TABLE t1(a varchar(60)) ENGINE INNODB' failed: <Unknown> (2013): Lost connection to server during query
```
The call stack is as follows:
```
server.dll!log_write_up_to()[log0log.cc:1100]
server.dll!trx_flush_log_if_needed()[trx0trx.cc:1283]
server.dll!trx_t::commit_in_memory()[trx0trx.cc:1504]
server.dll!trx_t::commit_low()[trx0trx.cc:1610]
server.dll!trx_t::commit_persist()[trx0trx.cc:1623]
server.dll!trx_t::commit()[trx0trx.cc:1634]
server.dll!dict_stats_save()[dict0stats.cc:3461]
server.dll!dict_stats_update()[dict0stats.cc:4153]
server.dll!ha_innobase::create()[ha_innodb.cc:13224]
```
I think that we must avoid specifying a `completion_callback` when `!flush && log_sys.is_mmap()` .

Reply:
cbd77f8685edbcaad3dc74994cc2963e261155a0 fixes this, and also ensures that the callback will actually be used when using regular system calls for log writes.